File: src/10.graph/31.200.ts
export function numIslands1(grid: string[][]): number {
    const m = grid.length;
    const n = grid[0].length;
    let ans = 0;
    const dfs = (i: number, j: number) => {
        if (i < 0 || i >= m || j < 0 || j >= n) {
            return;
        }

        if (grid[i][j] === "1") {
            grid[i][j] = "2";
            dfs(i - 1, j);
            dfs(i + 1, j);
            dfs(i, j - 1);
            dfs(i, j + 1);
        }
    };

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === "1") {
                ans++;
                dfs(i, j);
            }
        }
    }

    return ans;
}

/**
 * BFS
 * @param grid
 * @returns
 */
export function numIslands(grid: string[][]): number {
    const m = grid.length;
    const n = grid[0].length;
    let ans = 0;
    const bfs = (r: number, c: number) => {
        const stack: number[][] = [[r, c]];
        while (stack.length) {
            const cur = stack.pop();
            const [i, j] = cur;

            if (i < 0 || i >= m || j < 0 || j >= n) {
                continue;
            }

            if (grid[i][j] === "1") {
                grid[i][j] = "2";

                stack.push(
                    ...[
                        [i - 1, j],
                        [i + 1, j],
                        [i, j - 1],
                        [i, j + 1],
                    ]
                );
            }
        }
    };

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === "1") {
                ans++;
                bfs(i, j);
            }
        }
    }

    return ans;
}

File: src/10.graph/31.200.go
package graph

func NumIslands1(grid [][]byte) int {
	m := len(grid)
	n := len(grid[0])
	ans := 0
	var dfs func(i, j int)
	dfs = func(i, j int) {
		if i < 0 || i >= m || j < 0 || j >= n {
			return
		}

		if grid[i][j] == '1' {
			// 表示遍历过了
			grid[i][j] = '2'
			dfs(i-1, j)
			dfs(i+1, j)
			dfs(i, j-1)
			dfs(i, j+1)
			return
		}
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == '1' {
				ans++
				dfs(i, j)
			}
		}
	}

	return ans
}

// NumIslands  BFS 搜索
//
//	@param grid
//	@return int
func NumIslands(grid [][]byte) int {
	m := len(grid)
	n := len(grid[0])
	ans := 0

	bfs := func(i, j int) {
		stack := [][]int{{i, j}}

		for len(stack) > 0 {
			cur := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			r, c := cur[0], cur[1]

			if r < 0 || r >= m || c < 0 || c >= n {
				continue
			}

			if grid[r][c] != '1' {
				continue
			}

			grid[r][c] = '2'
			stack = append(stack, [][]int{
				{r - 1, c},
				{r + 1, c},
				{r, c - 1},
				{r, c + 1},
			}...)
		}
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == '1' {
				ans++
				bfs(i, j)
			}
		}
	}

	return ans
}
